### PR TITLE
mono_thread_info_set_is_async_context assert that setting to false is correct, i.e. no nesting

### PR DESCRIPTION
**Pull request by jaykrell**
mono_thread_info_set_is_async_context assert that setting to false is correct, i.e. no nesting
Possibly related to https://github.com/mono/mono/issues/9566

### CODE REVIEW

**Comment by BrzVlad** [comment]
Does this happen in the first place ? I don't see the use of recursiveness for this. Personally I'd rather have an assert just in case code will change.

**Comment by jaykrell** [comment]
I don't know.
This is where reading the code got me wrt that bugweek bug.
The pattern in the code seems suspicious and the pattern in the PR seems very safe.

**Comment by jaykrell** [comment]
It's not a recursive thing necessarily, it is what state is the thread in when we call this.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Checked Private Types Build

File: mono/utils/mono-threads.c [hunk 1]
```diff
@@ -1373,8 +1373,18 @@ mono_thread_info_set_is_async_context (gboolean async_context)
...
-	if (info)
+	if (info) {
+		// If this assert fails, that means there is recursion and/or
+		// concurrency, such that setting async_context to FALSE
+		// that all the callers do after this, is incorrect,
+		// and this should instead be incremented/decremented.
+		//
+		// As the value is only accessed via current(), that
+		// limits the case to recursion, but increment/decrement
+		// is still fast and correct and simple.
```
Comment (jaykrell) [inline mono/utils/mono-threads.c]: Ugh I edited the comment wrong. It used to mention save/set/restore..

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build OS X x64

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Coop Suspend

**Pull request merged**